cpp/bench/sg/dbscan.cu

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuML.hpp>
+#include <dbscan/dbscan.hpp>
+#include "dataset.h"
+#include "harness.h"
+
+namespace ML {
+namespace Bench {
+namespace dbscan {
+
+template <typename D>
+struct Params : public BlobsParams<D> {
```

Comment:
IMO, a more flexible approach could be to fully decouple the algorithms & parameters from the datasets. I really like how you've separated the concerns between the algorithm and the parameters, but it would be really nice if I could wire up, say, `BlobsParamsA` and run it with `DBScanParamsA` on `DBScan`. This would also allow us to reuse `BlobsParamsA` across different algorithms, specified at runtime. For example, a command-line application could be built that could allow us to specify the wiring between a dataset, algorithm, and its params. This information could even be stored in a json configuration file, or some other format.

Reply:
Very valid point. I suffered heavily due to my myopic design with this one, while updating to google bench! Check out the current code as it should be way cleaner than this.